oci/build-layer.py

```diff
@@ -0,0 +1,299 @@
+                index += 1
+
+            if index == 0:
+                mask = 0o777
+
+            while m := self.change_regex.match(mode_part, index):
+                index = m.end()
+                op = m.group('op')
+                change_str = m.group('change')
+
+                change = 0
+                for c in change_str:
+                    if c == 'r':
+                        change |= 0o444
+                    elif c == 'w':
+                        change |= 0o222
+                    elif c == 'x':
+                        change |= 0o111
+                    elif c == 'X':
+                        if is_dir or new_mode & 0o111 > 0:
+                            change |= 0o111
+                    elif c == 's':
+                        mask |= 0o7000
+                        change |= 0o4000 | 0o2000
+                    elif c == 't':
+                        mask |= 0o7000
+                        change |= 0o1000
+                    elif c == 'u':
+                        umode = new_mode & 0o700
+                        change = umode | umode >> 3 | umode >> 6
+                    elif c == 'g':
+                        gmode = new_mode & 0o070
+                        change = gmode << 3 | gmode | gmode >> 3
+                    elif c == 'o':
+                        omode = new_mode & 0o007
+                        change = omode << 6 | omode << 3 | omode
+
+                if op == '+':
+                    new_mode = new_mode | (mask & change)
+                elif op == '-':
+                    new_mode = new_mode & ~(mask & change)
+                else:
+                    new_mode = (new_mode & ~mask) | (mask & change)
+
+            if index != len(mode_part):
+                raise Exception('Invalid mode string: {}'.format(mode_part))
+
+        self.cache[key] = new_mode
+        return new_mode
+
+
+# This class assists with avoiding duplicate entries when copying in
+# directories into the layer. OCI layers disallow having duplicate entries in
+# the tar.
+class CopyFromSources:
+    def __init__(self, layer, path, info):
+        self.layer = layer
+        self.path = path
+        self.info = info
+        self.entries = set()
+
+    def add_external_entry(self, path):
+        self.entries.add(path)
+
+    def do_copy(self):
+        parent_uid = assert_number(self.info.get('uid', 0))
+        parent_gid = assert_number(self.info.get('gid', 0))
+
+        # copy in reverse order so that later sources take priority
+        for source in reversed(self.info['sources']):
+            uid = assert_number(source.get('uid', parent_uid))
+            gid = assert_number(source.get('gid', parent_gid))
+            mode = parse_mode(source.get('mode'))
+
+            def modifier(tarinfo):
+                tarinfo.mtime = self.layer.fixed_time
+                tarinfo.uid = uid
+                tarinfo.gid = gid
+                tarinfo.uname = str(uid)
+                tarinfo.gname = str(gid)
```

Comment:
By default, `tarfile.gettarinfo` will resolve these to actual names. We just set them to the numeric id to avoid depending on the host system.